Initialize number formatter country code from server-bootstrapped data

aduth opened the pull request:
## Proposed Changes

* Changes geolocation initialization to avoid a network request if the country code can be pulled from ~the current logged-in user~ the [server-populated cookie country code value](https://github.com/Automattic/wp-calypso/blob/329877ea2809a19aa0eae91b781eb7bd1adadefb/client/server/pages/index.js#L127-L129)

## Why are these changes being made?

* Avoid unnecessary network requests

## Testing Instructions

Verify requests to `/geo/` on boot ~only if logged out~:

1. Go to http://calypso.localhost:3000
2. Open Chrome DevTools Network tab
3. Refresh the page
4. Verify no requests to `/geo/` unless logged out

Note that this technically relies on cookies assigned through client-side code such as [the GDPR cookie geolocation behavior](https://github.com/Automattic/wp-calypso/blob/trunk/client/lib/analytics/utils/refresh-country-code-cookie-gdpr.js), since server forwarding of country code doesn't happen in local development. In production, the server should always forward the `country_code` to the client.

There's a couple potential edge-cases to consider here:

- We've changed country code initialization from asynchronous to synchronous on application load, so it's possible some code could depend on the previous asynchronous behavior (though this is unlikely)
- The country code from the cookie can take form of values not consistent with previous raw handling of `/geo/` endpoint, such as [assigning `'unknown'` value on failed requests](https://github.com/Automattic/wp-calypso/blob/329877ea2809a19aa0eae91b781eb7bd1adadefb/client/lib/analytics/utils/refresh-country-code-cookie-gdpr.js#L36). Part of the hope of consolidating approaches is to work toward a consistent, stable set of expected values for country codes.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

aduth committed:
Initialize currency from current user if available

chriskmnds reviewed (dismissed):
Hi. Can you please take a look at https://github.com/Automattic/wp-calypso/pull/102538, which extends `geoLocateCurrencySymbol` to accept a callback as one way of making it work with the number-formatters package? That PR is ready to go, just waiting on some underlying package configuration work to complete on the JP end.

p.s. As currency-formatting is the only use of this logic, and that is going to be replaced entirely by number-formatters package in the next couple of days, then I think this PR should build on top of that direction (hence my request for changes). But it could be done differently all around. Just let me know what you think.

jsnajdr commented:
I'd prefer to have the relevant code improved in two ways:

**Give more precise names to things:**
The `geolocateCurrencySymbol` function doesn't work with any "currency symbols" at all. It sets the _country code_. The `setGeoLocation` function doesn't set a geolocation. It sets, again, a country code. Geolocation is merely a method we use to guess the country code. And the country code is then used to guess a currency code. Let's name things precisely. Like `geolocateCountryCode` and `setCountryCode`.

**Separate the geolocation from the actual setting:**
There should be two functions:
- `geolocateCountryCode(): Promise<string>`. This function asynchronously does the `/geo` network request and returns the determined country code. It should not set the `i18n` internal property. In other words, it should not have any side effects. That makes it a helper function that does only one job and has no dependencies on the rest of the `i18n` library.
- `setCountryCode`, a library-specific setter for the `i18n` object or for the `numberFormatter`/`currencyFormatter` object.

There is no need for callback arguments, just return a `Promise` that is not a `Promise<void>`, but `Promise<CountryCode>`.

The Calypso initialization can the look like:
```js
const countryCode = currentUser
  ? currentUser.user_ip_country_code
  : await i18n.geolocateCountryCode();

i18n.setCountryCode( countryCode );
numberFormatter.setCountryCode( countryCode );
```
Note how the separation of labor between `geolocate` and `set` allows us to neatly use multiple detection methods (logged-in user vs client IP address) and multiple setters.

The current two PRs (#102538 and #102862) make steps in that direction, but still have some way to go.

aduth commented:
@jsnajdr @chriskmnds I'm trying to reconcile the suggestions with understanding where we're going with all this effort. Based on @chriskmnds's work in #102878, it seems like the idea is that the currency formatting will be moved out of `i18n-calypso` into the `number-formatters` package. At that point, `i18n-calypso` has no use for geolocation / country code. I wouldn't want to put a lot of effort into refining `i18n-calypso`'s geolocation APIs if we don't intend for it to need to manage geolocation.

On the other hand, I'm not sure I have a clear sense where exactly we would want the geolocation code to live at the end of all this work. Part of me thinks it's simple enough to inline into `boot/common.js`:

```ts
async function initializeFormatters( currentUser ) {
	const countryCode = currentUser
		? currentUser.user_ip_country_code
		: await fetch( '/geo' )
				.then( ( response ) => response.json() )
				.then( ( data ) => data.country_short );

	// defaultCalypsoI18n.setCountryCode( countryCode );
	numberFormatter.setCountryCode( countryCode );
}
```

However, I think we're also interested in deduplicating geolocation requests with [the `useGeolocation` hook](https://github.com/Automattic/wp-calypso/blob/trunk/client/data/geo/use-geolocation-query.ts) and generally moving toward reactive hooks, so maybe it makes sense to have this be handled in some shared layout component, like [`<CalypsoI18nProvider />`](https://github.com/Automattic/wp-calypso/blob/trunk/client/components/calypso-i18n-provider/index.tsx) ? (I'm not entirely convinced this is an "i18n" concern, but it's convenient and related enough)

```ts
const CalypsoI18nProvider: FunctionComponent< { i18n?: I18N; children?: React.ReactNode } > = ( {
	i18n = defaultCalypsoI18n,
	children,
} ) => {
	// ...
	const countryCode = useCountryCode();

	useEffect( () => {
		// i18n.setCountryCode( countryCode );
		numberFormatter.setCountryCode( countryCode );
	}, [ countryCode ] );
	// ...
};
```

aduth committed:
Sync formatter country codes via React Query hook

aduth commented:
Reverting this back to draft, but I pushed an exploration of ideas in my previous comment in ab17513 . Curious for some feedback on this direction and open to suggestions.

I think this does a few things well:

- Consolidates geolocation requests to the `useGeoLocation` React Query hook
- Incremental step toward reactive formatters
- Reduces the amount of work and code size of the boot code

A few complications I encountered:

- Being part of the React tree means we need to be conscious of a few things:
   - Is Redux state available? My first attempt at incorporating this in `CalypsoI18nProvider` failed because we render it outside the Redux context provider, although this could potentially change.
   - Is the layout being rendered server-side? We don't want to make geolocation requests server-side, though I'm curious if that's already happening since `geolocateCurrencySymbol` is in "common" boot code.

A few things I'm not so sure about:

- Adding more components. Could this we attach this into an existing component?
- Standalone "sync" component vs. context provider. I imagine the final state of a reactive currency formatter would likely be be reactive through React context, so we might need to do this eventually, but I'm conscious of adding more provider wrappers.

jsnajdr commented:
`useGeoLocationQuery`: I think this query hook was a mistake. Just compare it with how our React code accesses the "currrent user" and "current locale" variables. These are global object, singletons, initialized when the app boots. The React code just assumes they are there, accessible immediately, synchronously.

That's how country code should also work. Currently we call `useGeoLocationQuery` at 30 places, where each of them can in theory trigger the `/geo` REST request, and also has to deal with the async situation when the data is not yet available (`if ( ! geoData ) return null;`). Instead it should be initialized at one place. It's part of the "current user" data that we fetch at boot anyway. And in production it's bootstrapped by the server, for better performance.

We could also start bootstrapping the country code on the server for logged-out users. When you're logged out, the `/me` endpoint returns a 403 and we could piggy-back the country code to the error response:
```
"body": {
  "error": "authorization_required",
  "message": "An active access token must be used to query information about the current user.",
  "ip_country_code": "CZ"
}
```

There are many ways how we can actually expose the country code to React components. It can be stored in Redux, in a global variable, in a context provider. We can also initialize the React Query cache with the data if we want.

chriskmnds commented:
> `useGeoLocationQuery`: I think this query hook was a mistake. Just compare it with how our React code accesses the "currrent user" and "current locale" variables. These are global object, singletons, initialized when the app boots. The React code just assumes they are there, accessible immediately, synchronously.
> 

Perhaps a lot of that needs to change instead 🤔 
I'd expect the app to be reactive to changes, for example, in `/me/account`. Irrespective of how we initialise the context, in the background it could probably not rely on static data alone. Right now, if you change locale, you need to reload the page to take effect, while just refocusing would pick up the change. It's not something that users do often, though, obviously.

I'm sure we'll need to think at some point how packages used outside that consume these other packages will also need initialisation e.g. plans-grid imported in LandPack. Some sort of share thing / "wpcom i18n initialisation provider" might help, but little thought right now.

aduth commented:
Yeah I'm not sure I totally agree that a hooks-based approach is inappropriate for reactive changes to geolocation, although I grant that we're likely moving in a direction where the value is unchanging and always available to us through bootstrapped data. Another motivation here was to try to limit the amount of code in boot.

In the spirit of compromise and simplifying this iteration, maybe it's best to start with an approach similar the first example in my comment of https://github.com/Automattic/wp-calypso/pull/102862#issuecomment-2835690993, where we fetch and assign the country code in each of `i18n-calypso` and `number-formatters` from the boot code directly.

jsnajdr commented:
> I'd expect the app to be reactive to changes, for example, in `/me/account`. Irrespective of how we initialise the context, in the background it could probably not rely on static data alone. Right now, if you change locale, you need to reload the page to take effect, while just refocusing would pick up the change.

Calypso locale currently _is_ reactive. If you change the language or load new translations in the provider at the top of the React tree, an update will be fired and all the `useTranslate` and `useI18n` hook instances will rerender their components with the new data.

The point is that the locale updates are centralized at one place. The `useTranslate` hook is always merely a reader and an observer, there's no way how it could ever trigger any change in the locale or locale data. But it's still 100% reactive.

The `useGeoLocationQuery` hook is different: any instance of the hook can trigger a fetch and can do an update that all other instances will see. The fetching is decentralized, and therefore harder to manage: that's why we have so many `/geo` requests.

chriskmnds commented:
> > I'd expect the app to be reactive to changes, for example, in `/me/account`. Irrespective of how we initialise the context, in the background it could probably not rely on static data alone. Right now, if you change locale, you need to reload the page to take effect, while just refocusing would pick up the change.
> 
> Calypso locale currently _is_ reactive. If you change the language or load new translations in the provider at the top of the React tree, an update will be fired and all the `useTranslate` and `useI18n` hook instances will rerender their components with the new data.
> 
> The point is that the locale updates are centralized at one place. The `useTranslate` hook is always merely a reader and an observer, there's no way how it could ever trigger any change in the locale or locale data. But it's still 100% reactive.
> 
> The `useGeoLocationQuery` hook is different: any instance of the hook can trigger a fetch and can do an update that all other instances will see. The fetching is decentralized, and therefore harder to manage: that's why we have so many `/geo` requests.

> If you change the language 

Yes. I think my reference to being "reactive" is how they probe and react to endpoint changes e.g. 
- I go to `/me/account`, in another tab, change language
- I go back to `/stats/:site` in previous tab, nothing changes. I have to reload 

If I do the same with currency (which we pull from the plans endpoint via a Query with refocus):
- I go to `/wp-admin/network`, in another tab, change currency
- I go back to `/start/plans`, in another tab, changes propagate, no reload

----------

As for locale, I also don't like how you cannot create another instance of i18n to use with a different locale. Here's what I needed to do at one point to achieve that: https://github.com/Automattic/wp-calypso/pull/101444 (We needed to load translations for a notice in another language - they are currently doing it by sending the translated strings from the backend/endpoint).

chriskmnds commented:
Ok. I shipped the main PRs that complete the migration to number-formatters in wp-calypso. The two PRs left are:

1. https://github.com/Automattic/wp-calypso/pull/102878
2. https://github.com/Automattic/wp-calypso/pull/100559

I ran out of time, but I intend to ship the first one tomorrow, the second one next week. So we can already rebase this PR over #102878 to get the latest changes and deal only with populating `getLocation` in number-formatters package.

aduth committed:
Revert "Sync formatter country codes via React Query hook"

This reverts commit 54f51a7238d303478ec80e29a525a2553ad39b7c.

aduth committed:
Initialize country code in boot

aduth commented:
I rebased the branch and took a different direction to handle the country code as part of boot, assigning the result into `i18n-calypso` and `number-formatters` packages. Feedback welcome @chriskmnds @jsnajdr 

With this, I think we could remove `geolocateCurrencySymbol` from `i18n-calypso`, but maybe we want to do that separate?

chriskmnds commented on client/boot/geolocation.ts:
```diff
@@ -0,0 +1,31 @@
+		return;
+	}
+
+	setGeoLocation( countryCode );
```
This part is not reactive right? Meaning, if `countryCode` ends up coming async, it won't trigger a rerender per se / on its own. We update the singleton but nothing further happens. It's how it is now too, and not a big deal either. 

Or am I missing something?

chriskmnds replied:
> It's how it is now too, and not a big deal either.

Actually, not sure if that's true. Since now (currently) we update the i18n instance, it triggers other side effects, but nothing that's relevant to number-formatters. 

I'd rather we rebase over https://github.com/Automattic/wp-calypso/pull/102878, which removes `i18n.geoLocation` as no longer used. Just to make sure we test this isolated i.e. with the only effect on number-formatters.

aduth replied:
Yes, my expectation of the changes here was to maintain the current behavior of setting singleton values in `i18n-calypso` (for now, until removed) and the equivalent `setGeoLocation` in `number-formatters`. There's no reactivity being introduced here.

> > It's how it is now too, and not a big deal either.
> 
> Actually, not sure if that's true. Since now (currently) we update the i18n instance, it triggers other side effects, but nothing that's relevant to number-formatters.

As best I've been able to tell, the current geolocation assignment behavior in `i18n-calypso` _doesn't_ trigger any cascading effects ([doesn't call `this.stateOberver.emit( 'change' )`](https://github.com/Automattic/wp-calypso/blob/83b4f68b8c58d147550151e85ff659a2605aaa8e/packages/i18n-calypso/src/i18n.js#L198)), which might be problematic, but is how it's worked and is preserved here.

chriskmnds replied:
Thanks. Indeed only emits change for locale changing. In any case, `geoLocation` was also ported from `@automattic/format-currency`, which is how it works there too.

We'll probably need to explore introducing reactivity to number-formatters. I'm thinking encapsulating these with a hook like `const { formatCurrency, ... } = useNumberFormatters();`, then decide where the context will live. A brief exploration points me to a singleton store with `useSyncExternalStore`, which would make it the more "conventional" way outside of using context provider. But all that would be like a number-formatters v2.

chriskmnds commented:
> With this, I think we could remove geolocateCurrencySymbol from i18n-calypso, but maybe we want to do that separate? I can also include it here.

I think just do it here if it makes sense to test in one go.

There's no point in populating `i18n.geoLocation` either. Obsolete as of #102878 that will ship tomorrow.

aduth commented:
> > With this, I think we could remove geolocateCurrencySymbol from i18n-calypso, but maybe we want to do that separate? I can also include it here.
> 
> I think just do it here if it makes sense to test in one go.
> 
> There's no point in populating `i18n.geoLocation` either. Obsolete as of #102878 that will ship tomorrow.

In that case, I'll plan to rebase after that code is merged 👍 Hopefully that will simplify things a bit more.

aduth commented:
I stumbled upon another module where we're fetching and using country code:

[`client/lib/analytics/utils/refresh-country-code-cookie-gdpr.js`](https://github.com/Automattic/wp-calypso/blob/trunk/client/lib/analytics/utils/refresh-country-code-cookie-gdpr.js)

The module appears to be used in a number of different places, some of which could benefit from reactive country code. Or at the very least it would be nice to try to eliminate an extra network request and reuse the bootstrapped data in the same way we're using here.

aduth commented:
For logged-out users, I discovered that we already assign a cookie for `country_code` server-side using the forwarded `x-geoip-country-code` header, which I expect we could use:

https://github.com/Automattic/wp-calypso/blob/2f6ba361c31058968c3435fff03b011bd0342327/client/server/pages/index.js#L123-L125

e.g. for me visiting Calypso production while logged out (Incognito):

```ts
document.cookie.split(';').map((c) => c.trim().split('=')).find(([n]) => n === 'country_code')[1]
// 'US'
```

aduth committed:
Initialize country code from cookie

aduth commented:
I've rebased after the merge of #102878, and removed all of the geolocation functionality from the `i18n-calypso` package. I also migrated to using the cookie value as mentioned in [my previous comment](https://github.com/Automattic/wp-calypso/pull/102862#issuecomment-2845619332), which works for both logged-in and logged-out users and simplifies the code a bit.

jsnajdr commented:
> For logged-out users, I discovered that we already assign a cookie for `country_code` server-side

That's a great find, turns out that we are already always bootstrap the country code on the server! This was added few years ago by @sgomes when fixing SSR of the GDPR cookie banner. The need for the `/geo/` request should be very rare.

One thing that's a bit out of sync is the expiration of the `country_code` cookie. The server sets it as a session cookie, i.e., it lives only during the lifetime of the browser tab. On the other hand, the `refreshCountryCodeCookieGdpr` helper in analytics sets expiration to 6 hours. Not a serious problem at all, but still an inconsistency.

jsnajdr commented on client/boot/geolocation.ts:
```diff
@@ -0,0 +1,24 @@
+	const cookies = cookie.parse( document.cookie );
+	const countryCode = cookies.country_code ?? ( await fetchCountryCode() );
+	if ( countryCode ) {
+		setGeoLocation( countryCode );
```
The next thing to do here (in a followup PR) is to write the `countryCode` to Redux. And make the `getCurrentUserCountryCode` selector read the value from a dedicated Redux state slice, instead of a field in the `user` object. All usages of `useGeoLocationQuery` then can be migrated to a simple synchronous select from Redux.

There's only `useGeoLocationQuery` call (out of 13) that cares about something else than `country_code`. It's the `JetpackComFooter` that looks at `.region` to check if the region is California, and then it shows a legal notice relevant only for California.

The analytics GDPR code sets both `country_code` and `region` cookies. If the server could bootstrap both values, then the number of usages of `useGeoLocationQuery` could truly drop to zero. But the server bootstrapping won't be easy, it relies on special headers forwarded by Nginx to the Calypso server.

aduth replied:
> All usages of `useGeoLocationQuery` then can be migrated to a simple synchronous select from Redux.

It feels a little strange to intentionally revert a React Query hook to using Redux state, although I see the advantages (synchronous, clearer pathway to using bootstrapped-data, etc.).

At that point, should we even bother having a React Query hook, especially if we're already disabling all of its refresh behavior and initializing it from server-bootstrapped data? In other words, we're expecting the country code will always be there and never change, so we could just read from Redux state directly, or some other static context.

jsnajdr replied:
> It feels a little strange to intentionally revert a React Query hook to using Redux state

It's strange only when you think about it in terms of a "bad/old library" vs "good/new library". Considering "what't the right tool for this task", the country code is global shared state initialized at boot and Redux is a good tool to propagate the state (dependency-inject it) to various parts of the app.

> At that point, should we even bother having a React Query hook

The hook can disappear completely, it will just be a Redux selector that always returns the right value. We can also get rid of some of the async logic in components where we check if the query `data` is `null`.

jsnajdr reviewed (approved):
This looks great, thanks for taking this through several iterations :+1:

@chriskmnds does this also look good to you now?

chriskmnds reviewed (changes requested):
Just to make sure we don't break anything inadvertently. Should proxy matter? If I switch to a US host, reload `/start/plans`, it looks country code doesn't propagate properly. Should it?

**with this PR:**

<img width="1409" alt="Screenshot 2025-05-06 at 12 05 57 PM" src="https://github.com/user-attachments/assets/a46675a8-6177-4378-8fd7-ca58370832ff" />

**in production:**

<img width="1411" alt="Screenshot 2025-05-06 at 12 06 04 PM" src="https://github.com/user-attachments/assets/c9f8a557-5e92-4067-bdd2-160c45478444" />

But pls confirm on your end.

jsnajdr commented:
> If I switch to a US host, reload `/start/plans`, it looks country code doesn't propagate properly.

@chriskmnds This is a nice catch. I'd like to reproduce and debug the problem and need a little clarification:
- the "in production" screenshot is from `wordpress.com/start/plans` while proxied and logged in?
- the "with this PR" screenshot is from local dev server, `calypso.localhost:3000`? Or from something else, like `calypso.live` for the PR?
- are you proxied in both scenarios? Proxy definitely matters for geolocation, because your IP is different. For example, I'm physically in Pilsen (CZ) and unproxied my location is detected correctly. But when proxied, the other end of the proxy is physically either in Vienna (AT) or Frankfurt (DE). The detected country code can have three different values depending on proxy on/off and also on network connection (internet at home vs mobile data network).

Do you happen to how what is the actual geolocation (country code) in the two scenarios? US$ and $ are really two names for the same currency so maybe there's also something else at play.

chriskmnds commented:
> * the "in production" screenshot is from `wordpress.com/start/plans` while proxied and logged in?

yes. I use AutoProxxy - it's easy to switch hosts with it (use `proxy.atl....` for US).

> * the "with this PR" screenshot is from local dev server, `calypso.localhost:3000`? Or from something else, like `calypso.live` for the PR?

I tried both localhost and the live image to be sure actually. It's the same result, unless I am doing something wrong.

> are you proxied in both scenarios? 

Yes? I had the proxy enabled in both cases

> Do you happen to how what is the actual geolocation (country code) in the two scenarios? US$ and $ are really two names for the same currency so maybe there's also something else at play.

If the country code, so geolocation, is not within the US, then `formatCurrency` will prefix (or suffix in case of DE/EL) the currency values with `US` symbol e.g. `US$123` or `123 US$`, etc. This is to distinguish US dollars from Canadian etc. It's the reason we use this geolocation for pricing: See the notes in the [README](https://www.npmjs.com/package/@automattic/number-formatters?activeTab=readme#setgeolocation). So yeah, the difference in the images is the currency values (sorry if this wasn't clear 😬 ). The country code that's being checked is just "US" ([respective code](https://github.com/Automattic/jetpack/blob/trunk/projects/js-packages/number-formatters/src/number-format-currency/index.ts#L20))

p.s. worth cc'ing @sirbrillig , just to be sure he is onboard with changes that may affect prices (he owns the code ultimately) :-)

aduth commented:
I haven't had a chance to dig in, but my initial hunch is wondering if we're now assigning the country code so quickly (synchronously) in the app initialization that it might be affecting some code which had come to rely on the previous asynchronous initialization?

chriskmnds commented:
> I haven't had a chance to dig in, but my initial hunch is wondering if we're now assigning the country code so quickly (synchronously) in the app initialization that it might be affecting some code which had come to rely on the previous asynchronous initialization?

Well, I think I have concerns with relying on the static context (so the cookie or backend provided values) for anything beyond "initialisation". I meant to comment earlier, but I understand you are all doing a lot better thinking that I can. 

I think we need to ensure, in general, that further API/endpoint requests will reason with the same data that the frontend uses. So if geolocation (as an example really) becomes something that the endpoints use internally, then obviously we'll have clashes if the frontend uses something else for similar purposes (show some notice, render some price, etc.). Either always provide the frontend context to the endpoints then, or meet things in the middle by allowing to refresh these values as user interacts (so the Query hooks). I think the latter sounds more realistic.

Just my thinking on the moment.

aduth commented:
I was able to reproduce in calypso.live but not in local development. One thing I noticed in calypso.live where I could reproduce is that the cookie value for `region` and `country_code` was set to `unknown`. Trying to track this down, I notice we use `'unknown'` return values when the GDPR-related geo-data refresh fails:

https://github.com/Automattic/wp-calypso/blob/44f1d8824f84dfac457f852497c4df0b57a97db7/client/lib/analytics/utils/refresh-country-code-cookie-gdpr.js#L48-L51

These values are then [written into the cookie](https://github.com/Automattic/wp-calypso/blob/44f1d8824f84dfac457f852497c4df0b57a97db7/client/lib/analytics/utils/refresh-country-code-cookie-gdpr.js#L24-L25), which is what we're now using to initialize.

Which has me thinking a few follow-on points:

- Based on [the code you shared in how we're checking `US$`](https://github.com/Automattic/jetpack/blob/44a5d2e561f671040d689d305dedf9192ad69c11/projects/js-packages/number-formatters/src/number-format-currency/index.ts#L20), it's now understandable how a value of `'unknown'` would result in the prefix being added
   - Either (1) `'unknown'` should be an expected value and the `getCurrencyOverride` condition should be updated to check for it, or (2) `'unknown'` should not be what we assign into the cookie and instead we either don't assign it or assign an empty value like how we're otherwise using the country code
- The `requestGeoData` function where this is being produced is one another one of the geolocation fetchers that I'd like to consolidate into existing options, which may indirectly fix this issue
- It's unclear to me currently why this request is failing in the first place to be producing the `unknown` values

aduth commented:
> I think we need to ensure, in general, that further API/endpoint requests will reason with the same data that the frontend uses.

As far as I understand, the value of country code assigned in the server-side cookie (using [the forwarded `x-geoip-country-code` header](https://github.com/Automattic/wp-calypso/blob/a5bc60830d7cd473918f0c7dec6d814e37b62fa3/client/server/pages/index.js#L112)) should be the same as what we'd expect to be generated by the https://public-api.wordpress.com/geo/ endpoint, but this is worth confirming.

> meet things in the middle by allowing to refresh these values as user interacts (so the Query hooks)

We've actually moved a bit _further_ away from this in #102759 by avoiding refetches of geo data on the presumption that the geolocation is unlikely to change within the same page session, though I'd be curious your thoughts if you disagree with this premise. Granting that it's _technically possible_ someone might change geolocations, but [assessing](https://github.com/Automattic/wp-calypso/pull/102759#discussion_r2071452414) a low probability and low risk if the data were to turn stale.

The context of all this work wasn't stated in the original comment, but the goal is toward trying to reduce excess network requests like what was flagged in p58i-kIo-p2#comment-67529 .

aduth commented:
> * It's unclear to me currently why this request is failing in the first place to be producing the `unknown` values

I tracked it down to being an ad blocker extension blocking the `/geo` endpoint. I wasn't able to reproduce in local development because I've disabled the ad blocker on that host. It's likely that these endpoint calls are also blocked in production as well, but in production we don't fall back to `'unknown'` on a failed request like what's being done in [`refresh-country-code-cookie-gdpr.js`](https://github.com/Automattic/wp-calypso/blob/trunk/client/lib/analytics/utils/refresh-country-code-cookie-gdpr.js).

@chriskmnds Could you check to confirm if the issue persists after disabling your ad blocker and clearing cookies (at least `country_code` and `region`) for local development and/or calypso.live ?

aduth commented:
Two more concluding thoughts here:

- I expect that this behavior would not happen in production, since the [the forwarded `x-geoip-country-code` header](https://github.com/Automattic/wp-calypso/blob/a5bc60830d7cd473918f0c7dec6d814e37b62fa3/client/server/pages/index.js#L112) would be provided in production environments and [short-circuit the client-side fetch](https://github.com/Automattic/wp-calypso/blob/a5bc60830d7cd473918f0c7dec6d814e37b62fa3/client/lib/analytics/utils/refresh-country-code-cookie-gdpr.js#L13), but the `x-geoip-country-code` value is not provided in local development or in calypso.live, as far as I'm aware.
- Based on how we [only set the cookie server-side if it's a truthy value](https://github.com/Automattic/wp-calypso/blob/a5bc60830d7cd473918f0c7dec6d814e37b62fa3/client/server/pages/index.js#L123-L124), I think we should similarly not set the cookie client-side on a failed request. We may need a compensating behavior to avoid repeatedly fetching client-side, like a separate "fetched recently" cookie.

chriskmnds commented:
I can return to test here later. I'm a little lost in other work right now.

@aduth I think we should update the description to be a little more thorough on the changes. Also with some examples of how we confirm we are not affecting existing behavior: The plans grid with examples derived from both `/start/plans` and `/plans/:site` are good candidates as they bring a lot of things together (derive prices from endpoints, use geolocation for the US prefix, etc.). And if this PR needs the changes from https://github.com/Automattic/wp-calypso/pull/103195 then rebase over that and change parent here?

aduth commented:
Thanks @chriskmnds . No rush, and yes, I think the direction of #103195 will influence how to proceed here, so this pull request is very much not ready to merge regardless.

I'll plan to update the pull request description as well. Truthfully I hoped this change would be relatively unimpactful, hence the brief original description. This discussion has surfaced a number of inconsistencies throughout the application in the what values we expect for "country codes" or "regions". Some code expects strings, others nullish values, some `'unknown'`, and now with #103195 also surfacing `-` as a potential API value (for `region`). These are painful to be discovering in this work, though hopefully it's valuable in the long-term to be reconciling toward a consistent set of expected values.

jsnajdr commented:
#103195 is now merged, and after rebasing this PR on top, the currency formatting issue on the plans page should be fixed. Now we're no longer overwriting valid `country_code` cookie with an `unknown` value.

aduth commented:
I'll need to rebase and resolve conflicts, and check that @jsnajdr suffices to fix the original issue. Hoping to have that taken care of by the end of today.

aduth commented:
@chriskmnds To your broader point, I don't think it's as much that geolocation is strictly unchangeable, but that we lack meaningful **signals** to know when changes occur. I think we're trying to find a sweet spot on the spectrum of real-time vs. static, where we were previously polling quite regularly for updates.

aduth committed:
Add check for unknown countryCode value

aduth commented:
I force-pushed a rebase, which should incorporate the improvements from #103195. While I don't think it's ever likely to be triggered in production, I added a check for the `'unknown'` value in 808c76b, since technically this value can be made the cookie value in some circumstances (no server-populated cookie + failed geolocation request).

I updated the original pull request to reflect the changes in direction and edge-cases to look out for in testing. I'm having a hard time tracking down exactly how the server `x-geoip-country-code` header is set to [confirm](https://github.com/Automattic/wp-calypso/pull/102862#issuecomment-2855441785) that it's guaranteed to be in the same format as what's given through the `/geo` endpoint, but I'll keep looking.

I also want to add a bit more test coverage, but that shouldn't block testing.

aduth commented:
> What would a "meaningful signal" look like? 🤔

In a perfect world we could use a "push" signal, something like the browser's [Geolocation `watchPosition`](https://developer.mozilla.org/en-US/docs/Web/API/Geolocation/watchPosition). But I expect this would come with a permissions dialog that would cause alarm for users.

Practically speaking I think we're just going to have to make due with "pull" signals like polling or fetching on application load, as we've done historically.

aduth committed:
Add test coverage for geolocation bootstrapping

chriskmnds commented:
Hey! Apologies for the delay, and thanks for waiting! 
I tested again. I see the same difference/issue as before when switching proxies.
I tested in Incognito/Chrome Staging and Calypso Live in separate tabs. Exact steps:

- Proxy through a US host
- Go to `/start/plans` in separate tabs for Staging and Calypso Live (log in and return, currency on account is `$`)

| Staging | Calypso Live |
|--------|--------|
| <img width="1663" alt="Screenshot 2025-05-15 at 7 25 44 PM" src="https://github.com/user-attachments/assets/1a84b717-dc3d-4927-9a29-6a4a49dc52cf" /> | <img width="1648" alt="Screenshot 2025-05-15 at 7 26 07 PM" src="https://github.com/user-attachments/assets/ec277789-c7c5-4121-8076-15e06d4b92aa" /> | 

- Switch proxy to non-US
- Reload each tab

| Staging | Calypso Live (missing "US" prefix) |
|--------|--------|
| <img width="1642" alt="Screenshot 2025-05-15 at 7 26 57 PM" src="https://github.com/user-attachments/assets/505f28f2-e06f-43fe-ab7a-014ab7b865fc" /> | <img width="1646" alt="Screenshot 2025-05-15 at 7 27 11 PM" src="https://github.com/user-attachments/assets/330cbbf6-0a55-4be3-8c86-484b6fd2b884" /> | 

I marked the difference in the last media in a red box.

(Never mind the `Save 123%`, you'll probably not see that on your end. It's just part of an experiment I'm assigned to)

-----------

I think we need more testing. It seems to be a regression, but only if proxy matters (switching proxies)? Please confirm, perhaps with @sirbrillig as well.

sirbrillig commented:
> I think we need more testing. It seems to be a regression, but only if proxy matters (switching proxies)? Please confirm, perhaps with @sirbrillig as well.

Ideally if the user is coming from a non-US IP, we'd display the `US$` symbol. I imagine the scenario you tested might be a problem if a user is in Argentina and normally uses a US-based proxy but then turns it off to make a purchase and sees the screenshot on the right, mistakenly thinking that `$25` means `ARS$25`, which is quite a lot different than what it actually means, which is `US$25`. It's an edge case, perhaps, but it would be nice to avoid it if possible, especially if caching would cause the same problem when they reached checkout.

aduth commented:
I _think_ what could be happening is that staging and production will always assign a fresh cookie using [the server-forwarded `x-geoip-country-code` header](https://github.com/Automattic/wp-calypso/blob/a40795bc0bca860c173f8fd6c0a92bf9c64a1d50/client/server/pages/index.js#L112), but this header isn't available in calypso.live or in local development, so it relies instead on the cookie fetched in the client which has a 6 hour duration. So when you switched proxies the cookie wasn't refreshed.

But I'll double-check this is what's happening, and whether there's anything we can or should do about it.

You might actually encounter the same behavior on any other branch (including `trunk`) in local development or calypso.live

aduth commented:
>I don't see the issue there in my testing. It reacts properly to proxy changes. Perhaps I misunderstood, though - other branches rely entirely on the geo endpoint right now. 🤔

Ah yeah, you're right 🤦 I guess what would be more accurate to say is that I don't think you'd observe those same issues appearing in staging or production once this is merged.

In any case, I'll check to confirm, and see if we can do a better job of bringing the local and production behavior in sync. It would be nice if we could have [the `x-geoip-country-code` header forwarding](https://github.com/Automattic/wp-calypso/blob/116d185e3e88142a72cfbfe9c00f7720ae6b101a/client/server/pages/index.js#L112) work the same way locally as it does in production.

chriskmnds commented:
Thanks @jsnajdr for digging into this. Please feel free to proceed with this PR since we got to the bottom of it. 

> Calling `numberFormatters.setGeoLocation` doesn't cause a rerender. 

> 3. Make the currency formatter somehow reactive. The `getCurrencyObject` API wouldn't be static, but it would a return value of a React hook (`useNumberFormatter`?) that triggers a rerender when its internal state (`localeState` and `geoLocationState`) changes. Then the pricing page would display `$` for a moment but then quickly switch to the correct `US$`.

Yes. I created an issue recently in Linear to address this very thing: [i18n: number-formatters v2. Ensure "reactivity", make it reactive](https://linear.app/a8c/issue/M4R73CH-809/i18n-number-formatters-v2-ensure-reactivity-make-it-reactive)

It's not something that @Automattic/martech or @Automattic/i18n will be working on any time soon, so it's not prioritised right now AFAIK. I'd be happy to work on it, no doubt, just not within team scope. Please feel free to move it under team Calypso if it's something you are interested in.

aduth commented:
I created a separate pull request at #103597 to better simulate the expected production experience in local development.
